Change report modal to include category selection in web UI

Gargron opened the pull request:
A redesign of the report dialog in the web UI which integrates the report category functionality. You select the reason for reporting at the start, and are led through multiple screens depending on your choices. The violation category leads to an additional screen where you can check off which server rules you think were violated. As before, there is a screen for attaching statuses and a screen for writing additional comments.  The report dialog now displays a thank you message at the end instead of automatically closing. The "I don't like it" choice at the beginning leads right to the "thank you" screen, which presents options such as muting and blocking the account.

|Step 1|Step 2|Step 3|Step 4|Step 5|
|-|-|-|-|-|
|![image](https://user-images.githubusercontent.com/184731/154776340-8e5aa229-8711-493f-ba91-958976f18ec9.png)|![image](https://user-images.githubusercontent.com/184731/154776502-ab844e30-4e18-45e4-97af-0e558fba9621.png)|![image](https://user-images.githubusercontent.com/184731/154776527-9d213c03-b937-4cbe-9098-1ce8f2b469bd.png)|![image](https://user-images.githubusercontent.com/184731/154776545-c5428d23-dffb-4909-9a76-0ef1dda266ba.png)|![image](https://user-images.githubusercontent.com/184731/154776567-85c45932-2fb1-45b4-aee6-c1db478faa9d.png)|

___

Funded through the BMBF (BMBF Förderkennzeichen: 01IS21S29)

Gargron committed:
Change report modal to include category selection in web UI

ClearlyClaire commented:
Overall it looks nice and welcoming, and the addition of a last screen is useful, however:
- it's a lot more clicks to report a single user, which may be a problem when facing e.g. harassment from multiple accounts
- what's the difference between “I don't like it” and “It's something else”?
- the thanks screen should probably check whether the person is already blocked or muted, in which case it would alter the message/suggestions; otherwise I'm afraid it would only be slightly confusing “uh? didn't I block this user already?”

ClearlyClaire commented:
Reading the code (which I haven't fully reviewed yet), my understanding is that “I don't like it” skips to the last screen without actually submitting a report. If that's true, then that last screen's text for this case is seriously misleading.

ClearlyClaire commented:
If the screen does not have enough vertical space (such as when using some phones in landscape mode), the modal is unusable:
![Screen Shot 2022-02-19 at 14 40 26](https://user-images.githubusercontent.com/384364/154803303-5a7d6826-3763-4a61-a1f1-5b21f3ffbd14.png)
![Screen Shot 2022-02-19 at 14 41 12](https://user-images.githubusercontent.com/384364/154803281-ab73fc64-c120-4bed-a3a9-40d3f0e96c55.png)
![Screen Shot 2022-02-19 at 14 41 25](https://user-images.githubusercontent.com/384364/154803288-331df62a-2e83-44c5-8fc0-fc4f9c326519.png)

ClearlyClaire commented:
Also, the forwarded report won't include the category and rules, so if the report relies on that, it may be of lesser quality. Not much that can be done about that (I mean, we could add a property or two for that, but ultimately, local rules may not mean much to a remote user's admin), but maybe adding a bit to the “The account is from another server. Send anonymized copy of the report there as well?” text may help. For instance “The forwarded report will include statuses and additional comments, but not the selected category”.

Gargron committed:
Various fixes and improvements

- Change thank you text to be different based on category
- Change starting headline to be different for account and status reports
- Change toggle components to have a checkmark when checked
- Fix report dialog being cut off on small screens
- Fix thank you screen offering mute or block if already muted or blocked
- Refactor toggle components in report dialog into one component

Gargron commented:
Addressed points (except federation of category, though I feel like it could actually be fairly easily added)

ClearlyClaire commented on app/javascript/mastodon/features/report/thanks.js:
```diff
@@ -0,0 +1,78 @@
+    return (
+      <React.Fragment>
+        <h3 className='report-dialog-modal__title'>{submitted ? <FormattedMessage id='report.thanks.title_actionable' defaultMessage="Thanks for reporting, we'll look into this." /> : <FormattedMessage id='report.thanks.title' defaultMessage='Thank you for reporting.' />}</h3>
+        <p className='report-dialog-modal__lead'>{submitted ? <FormattedMessage id='report.thanks.take_action_actionable' defaultMessage='While we review this, you can take action against @{name}:' values={{ name: account.get('username') }} /> : <FormattedMessage id='report.thanks.take_action' defaultMessage='Here are your options for controlling what you see:' />}</p>
+
+        {options.reduce((prev, curr) => [prev, <hr />, curr])}
```
Options might be empty. Furthermore, I'm not sure it makes sense to suggest muting an account when it's already blocked.

Gargron replied:
What do you think it should look like?

ClearlyClaire replied:
That's a good question. I guess something like “Since you have blocked <username>, you should not be seeing any of their toots going forward.”, not something long, just a reminder that you have blocked them, that may be useful especially in the “Block & Report” shortcut, as opposed to the confusing behavior of offering to block them when they were already blocked.

Gargron replied:
Alternative I can think of is showing the mute/block options but have the buttons reflect already-blocked state.

ClearlyClaire replied:
We could, but I think it would still be slightly confusing. Less than the initial version though.

ClearlyClaire replied:
Btw, muting or blocking from the last page of the modal resets the report modal altogether, while I guess it should just close it?

Gargron commented on app/javascript/mastodon/features/report/category.js:
```diff
@@ -0,0 +1,93 @@
+import Option from './components/option';
+
+const messages = defineMessages({
+  dislike: { id: 'report.reasons.dislike', defaultMessage: 'I don\'t like it' },
+  dislike_description: { id: 'report.reasons.dislike_description', defaultMessage: 'It is not something you want to see' },
+  spam: { id: 'report.reasons.spam', defaultMessage: 'It\'s spam' },
+  spam_description: { id: 'report.reasons.spam_description', defaultMessage: 'Malicious links, fake engagement, or repetetive replies' },
+  violation: { id: 'report.reasons.violation', defaultMessage: 'It violates server rules' },
+  violation_description: { id: 'report.reasons.violation_description', defaultMessage: 'You are aware that it breaks specific rules' },
+  other: { id: 'report.reasons.other', defaultMessage: 'It\'s something else' },
```
CodeClimate complains about using double quotes even if the string contains single quotes.

ClearlyClaire replied:
I think thay may warrant a change of CodeClimate's configuration, but it's not a big deal anyway.

Gargron commented on app/javascript/mastodon/features/report/thanks.js:
```diff
@@ -0,0 +1,78 @@
+
+    return (
+      <React.Fragment>
+        <h3 className='report-dialog-modal__title'>{submitted ? <FormattedMessage id='report.thanks.title_actionable' defaultMessage="Thanks for reporting, we'll look into this." /> : <FormattedMessage id='report.thanks.title' defaultMessage='Thank you for reporting.' />}</h3>
```
Okay, I'll bring it up with the designer.

Gargron commented on app/javascript/mastodon/features/report/components/option.js:
```diff
@@ -0,0 +1,60 @@
+
+    return (
+      <label className='dialog-option poll__option selectable'>
+        <input type={multiple ? 'checkbox' : 'radio'} name={name} value={value} checked={checked} onChange={this.handleChange} />
+
+        <span
+          className={classNames('poll__input', { active: checked })}
```
I understand that for consistency with polls and system input designs, multiple choice boxes should be square, but in the iOS design we used circles for both types and I wanted to give it a go. I do not believe there should be actual confusion over whether or not you can select multiple, since the text says "Select all that apply" and it's otherwise pretty obvious from context.

ClearlyClaire replied:
> but in the iOS design we used circles for both types and I wanted to give it a go

but… why… circles are pretty much universally for radio buttons and squares for checkboxes…

Gargron committed:
Change wording on final screen

Gargron commented:
Changed wording on the thank you screen when "I don't like this" was selected:

![image](https://user-images.githubusercontent.com/184731/155234037-d26101bb-7dd1-4e3b-9408-a982ee970bfb.png)

Added option to unfollow when the reported account is followed. Made it so the mute/block buttons say "Muted" and "Blocked" when the account is already muted or blocked. Made all actions close the modal. Chose to skip the mute dialog because I think opening another dialog right after this one is not good.

ClearlyClaire commented:
That the “Muted” and “Blocked” buttons are clickable is slightly weird but is not a big deal. Same with the buttons closing the modal without any kind of feedback. I think those things can be improved upon later.

I still don't understand why you would use round buttons for checkboxes, in opposition to design conventions used everywhere in Mastodon and outside. Since you're bringing up the iOS app, let me quote Apple's human interface guidelines (for MacOS, as for iOS it doesn't seem to recommend checkboxes or radio buttons at all, but switch buttons and pickers or table views):

> **Use a set of checkboxes when the user needs the ability to select multiple options at the same time.** Except in a rare case where a mixed state is needed, radio buttons should be mutually exclusive.

https://developer.apple.com/design/human-interface-guidelines/macos/buttons/radio-buttons

Gargron committed:
Change checkboxes to be square when multiple options are possible

Gargron commented:
Changed because I don't want border-radius to block an important feature, but personally I liked the round ones more

shleeable commented:
Excellent PR. Thanks 👍

tribela commented:
![image](https://user-images.githubusercontent.com/5047683/155467446-93260d71-c051-42b3-9b9a-4c94e1036d8f.png)
Publish button got bigger text.
was it intended?

Gargron commented:
Yes, I adjusted button styles. I might make the text slightly smaller again though, it doesn't look quite right.
